### ISSUE TITLE
TestBackup timeouts

### ISSUE THREAD

**Issue by legrostdg** [restic/restic#1834]
Hi,

I'm trying to build v0.9.0 for Debian. I got a timeout for TestBackup on my laptop (i5-2540M and 8G RAM). It may be related to https://github.com/restic/restic/issues/1656, but as the timeouts happens on a "decent" machine (as opposed to the very restricted environment used to build architectures less popular than amd64 or i386 in the Debian archives), I thought it may be another problem. It seems to be the only issue preventing me from uploading v0.9.0 to unstable.

```
=== RUN   TestBackup
SIGQUIT: quit
PC=0x45d7e1 m=0 sigcode=0

goroutine 0 [idle]:
runtime.futex(0xf91b68, 0x0, 0x0, 0x0, 0x0, 0xc420279800, 0x0, 0x0, 0x7ffde2faee48, 0x410edb, ...)
	/usr/lib/go-1.10/src/runtime/sys_linux_amd64.s:530 +0x21
runtime.futexsleep(0xf91b68, 0xc400000000, 0xffffffffffffffff)
	/usr/lib/go-1.10/src/runtime/os_linux.go:45 +0x4b
runtime.notesleep(0xf91b68)
	/usr/lib/go-1.10/src/runtime/lock_futex.go:151 +0x9b
runtime.stopm()
	/usr/lib/go-1.10/src/runtime/proc.go:1952 +0xe5
runtime.findrunnable(0xc420034500, 0x0)
	/usr/lib/go-1.10/src/runtime/proc.go:2415 +0x50c
runtime.schedule()
	/usr/lib/go-1.10/src/runtime/proc.go:2541 +0x13b
runtime.park_m(0xc42041e600)
	/usr/lib/go-1.10/src/runtime/proc.go:2604 +0xb6
runtime.mcall(0x0)
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:351 +0x5b

goroutine 1 [chan receive, 9 minutes]:
testing.(*T).Run(0xc4203331d0, 0xba5f63, 0xa, 0xbd3bd0, 0x481901)
	/usr/lib/go-1.10/src/testing/testing.go:825 +0x301
testing.runTests.func1(0xc4202680f0)
	/usr/lib/go-1.10/src/testing/testing.go:1063 +0x64
testing.tRunner(0xc4202680f0, 0xc420203df8)
	/usr/lib/go-1.10/src/testing/testing.go:777 +0xd0
testing.runTests(0xc42025db20, 0xf898a0, 0x1b, 0x1b, 0x411ef9)
	/usr/lib/go-1.10/src/testing/testing.go:1061 +0x2c4
testing.(*M).Run(0xc42013e500, 0x0)
	/usr/lib/go-1.10/src/testing/testing.go:978 +0x171
main.main()
	_testmain.go:94 +0x151

goroutine 5 [syscall, 9 minutes]:
os/signal.signal_recv(0x0)
	/usr/lib/go-1.10/src/runtime/sigqueue.go:139 +0xa6
os/signal.loop()
	/usr/lib/go-1.10/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.0
	/usr/lib/go-1.10/src/os/signal/signal_unix.go:28 +0x41

goroutine 6 [chan receive, 9 minutes]:
github.com/restic/restic/internal/restic.init.0.func1.1()
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/restic/lock.go:257 +0xef
created by github.com/restic/restic/internal/restic.init.0.func1
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/restic/lock.go:254 +0x35

goroutine 7 [select, 9 minutes, locked to thread]:
runtime.gopark(0xbd4b90, 0x0, 0xba2cbc, 0x6, 0x18, 0x1)
	/usr/lib/go-1.10/src/runtime/proc.go:291 +0x11a
runtime.selectgo(0xc420046f50, 0xc42002a300)
	/usr/lib/go-1.10/src/runtime/select.go:392 +0xe50
runtime.ensureSigM.func1()
	/usr/lib/go-1.10/src/runtime/signal_unix.go:549 +0x1f4
runtime.goexit()
	/usr/lib/go-1.10/src/runtime/asm_amd64.s:2361 +0x1

goroutine 18 [chan receive, 9 minutes]:
github.com/restic/restic/internal/restic.init.2.func1(0xc42002a240)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/restic/progress_unix.go:17 +0xa3
created by github.com/restic/restic/internal/restic.init.2
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/restic/progress_unix.go:16 +0xab

goroutine 19 [chan receive, 9 minutes]:
github.com/restic/restic/cmd/restic.CleanupHandler(0xc42013a240)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/cmd/restic/cleanup.go:62 +0x5d
created by github.com/restic/restic/cmd/restic.init.0
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/cmd/restic/cleanup.go:24 +0x6d

goroutine 63 [chan receive, 9 minutes]:
gopkg.in/tomb%2ev2.(*Tomb).Wait(0xc422cfc690, 0x0, 0x0)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/gopkg.in/tomb.v2/tomb.go:126 +0x4a
github.com/restic/restic/cmd/restic.runBackup(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/cmd/restic/cmd_backup.go:480 +0x114e
github.com/restic/restic/cmd/restic.testRunBackup(0xc3cf40, 0xc4203331d0, 0xc420368e70, 0x1a, 0xc4201408e0, 0x1, 0x1, 0x0, 0x0, 0x0, ...)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/cmd/restic/integration_test.go:72 +0x384
github.com/restic/restic/cmd/restic.TestBackup(0xc4203331d0)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/cmd/restic/integration_test.go:254 +0x6d5
testing.tRunner(0xc4203331d0, 0xbd3bd0)
	/usr/lib/go-1.10/src/testing/testing.go:777 +0xd0
created by testing.(*T).Run
	/usr/lib/go-1.10/src/testing/testing.go:824 +0x2e0

goroutine 8 [select, 5 minutes]:
github.com/restic/restic/cmd/restic.refreshLocks(0xf90500, 0xc42002a3c0)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/cmd/restic/lock.go:72 +0x221
created by github.com/restic/restic/cmd/restic.lockRepository
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/cmd/restic/lock.go:49 +0x373

goroutine 283 [select, 9 minutes]:
github.com/restic/restic/internal/ui/termstatus.(*Terminal).runWithoutStatus(0xc420077140, 0xc360e0, 0xc4200750c0)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/ui/termstatus/status.go:191 +0x16f
github.com/restic/restic/internal/ui/termstatus.(*Terminal).Run(0xc420077140, 0xc360e0, 0xc4200750c0)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/ui/termstatus/status.go:87 +0xa2
github.com/restic/restic/cmd/restic.testRunBackup.func1(0x8, 0xbd4cf0)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/cmd/restic/integration_test.go:63 +0x3c
golang.org/x/sync/errgroup.(*Group).Go.func1(0xc420075100, 0xc422c4d080)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/golang.org/x/sync/errgroup/errgroup.go:58 +0x57
created by golang.org/x/sync/errgroup.(*Group).Go
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/golang.org/x/sync/errgroup/errgroup.go:55 +0x66

goroutine 419 [chan send, 9 minutes]:
github.com/restic/restic/internal/ui.(*Backup).ReportTotal(0xc422cf7860, 0xc422e1de60, 0x2c, 0x32, 0x0, 0x0, 0xc8000)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/ui/backup.go:336 +0x86
github.com/restic/restic/internal/ui.(*Backup).ReportTotal-fm(0xc422e1de60, 0x2c, 0x32, 0x0, 0x0, 0xc8000)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/cmd/restic/cmd_backup.go:428 +0x66
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc422ce1e00, 0xc360e0, 0xc420075500, 0x32, 0x0, 0x0, 0xc8000, 0xc422e1de60, 0x2c, 0x31, ...)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/archiver/scanner.go:110 +0x33a
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc422ce1e00, 0xc360e0, 0xc420075500, 0x0, 0x0, 0x0, 0x0, 0xc422e1cb70, 0x29, 0x27, ...)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/archiver/scanner.go:100 +0x5cf
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc422ce1e00, 0xc360e0, 0xc420075500, 0x0, 0x0, 0x0, 0x0, 0xc422e1cae0, 0x27, 0x25, ...)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/archiver/scanner.go:100 +0x5cf
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc422ce1e00, 0xc360e0, 0xc420075500, 0x0, 0x0, 0x0, 0x0, 0xc422e1ca50, 0x25, 0xc422e1c9c0, ...)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/archiver/scanner.go:100 +0x5cf
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc422ce1e00, 0xc360e0, 0xc420075500, 0x0, 0x0, 0x0, 0x0, 0xc422e1c9c0, 0x23, 0xbd4cd0, ...)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/archiver/scanner.go:100 +0x5cf
github.com/restic/restic/internal/archiver.(*Scanner).Scan(0xc422ce1e00, 0xc360e0, 0xc420075500, 0xc420140990, 0x1, 0x1, 0x4, 0xc420140c10)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/internal/archiver/scanner.go:51 +0x168
github.com/restic/restic/cmd/restic.runBackup.func4(0xbd4a20, 0x0)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/github.com/restic/restic/cmd/restic/cmd_backup.go:431 +0x9a
gopkg.in/tomb%2ev2.(*Tomb).run(0xc422cfc690, 0xc420075740)
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/gopkg.in/tomb.v2/tomb.go:163 +0x2b
created by gopkg.in/tomb%2ev2.(*Tomb).Go
	/<<BUILDDIR>>/restic-0.9.0+ds/obj-x86_64-linux-gnu/src/gopkg.in/tomb.v2/tomb.go:159 +0xb9

rax    0xca
rbx    0xf91a20
rcx    0x45d7e3
rdx    0x0
rdi    0xf91b68
rsi    0x0
rbp    0x7ffde2faee10
rsp    0x7ffde2faedc8
r8     0x0
r9     0x0
r10    0x0
r11    0x286
r12    0xcce498
r13    0x0
r14    0x0
r15    0x0
rip    0x45d7e1
rflags 0x286
cs     0x33
fs     0x0
gs     0x0
*** Test killed with quit: ran too long (10m0s).
FAIL	github.com/restic/restic/cmd/restic	600.012s
```

**Comment by fd0** [comment]
That's very odd, I'm having a look. How did you run the test, how can I reproduce this issue?

The test is very fast for me (thinkpad t460p, i5):
```
$ cd cmd/restic; go test -run 'TestBackup$'
PASS
ok      github.com/restic/restic/cmd/restic     0.183s
```

All tests within `cmd/restic` run in ~6s on my machine:
```
$ go test  
[...]
ok      github.com/restic/restic/cmd/restic     5.697s
```

**Comment by fd0** [comment]
It looks like a deadlock to me, not a performance issue. Hm. In #1656 we reduced the system resources needed to run the tests considerably.

**Comment by fd0** [comment]
I've found the bug: When the archiver is done before the scan process, restic deadlocks.

**Comment by fd0** [comment]
So, what do we do once #1835 is merged? I can see two options:
 * Release restic 0.9.1
 * Just add the one commit as a patch on top of 0.9.0 and package that for Debian

What's your preference?

**Comment by legrostdg** [comment]
It would be easier for me if you released 0.9.1. Thanks for the fast fix!

**Comment by fd0** [comment]
Ok, then we'll do that.

**Comment by fd0** [comment]
I've just released 0.9.1, please go ahead and report other issues :P
